### PR TITLE
Issue 2011: Stream truncation integration and system test

### LINKED ISSUE #2011: Stream truncation integration test
**Problem description**
Add more integration tests for truncation. 
At least these two:
 1. get a valid stream cut with offset > 0.
 validate truncation within a segment
2. have an existing reader reading from non truncated segment and then truncate the segment.
 verify that reader gets appropriate response and handles it successfully. 
3. have an existing reader reading from a segment that gets deleted while reader is reading from it. 

**Problem location**
test/integration/test/EndToEndTruncationTest

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**
Implemented integration tests and a system test for stream truncation.

- 4 new integration test cases have been developed in EndToEndTruncationTest.java.

- A new system test has been developed in OffsetTruncationTest.java.

**Purpose of the change**
Fixes #2011.

**What the code does**
The new integration tests cover the requirements of #2011:

> Get a valid stream cut with offset > 0 and validate truncation within a segment.

This requirement is validated with testSimpleOffsetTruncation in a simple scenario with a single-segment Stream and 1 reader/1 writer. A more elaborated scenario is defined in testParallelSegmentOffsetTruncation, as we can define multiple parallel segments and readers to validate the correctness of offset truncation.

> Have an existing reader reading from non-truncated segment and then truncate the segment. Verify that reader gets appropriate response and handles it successfully.

This situation is tested in testSegmentTruncationWhileReading.

> Have an existing reader reading from a segment that gets deleted while reader is reading from it.

This situation is tested in testDeleteStreamWhileReading.

Apart from integration tests, the system test contributed fulfills the first requirement in a distributed setting taking into account streams with parallel segments. Similar tests can be developed to cover other issues like #2438.

**How to verify it**

Execute new integration tests.
Execute the new system test (cluster).

### CODE REVIEW

**Comment by RaulGracia** [comment]
@fpj Now the changes are up to date and pass all checks.

File: test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndTruncationTest.java [hunk 1]
```diff
@@ -151,11 +164,283 @@ public void testTruncation() throws Exception {
...
+            log.info("Writing event: {} to stream {}", i, streamName);
+        }
+    }
+
+    private void writeDummyEvents(ClientFactory clientFactory, String streamName, int totalEvents) {
+        writeDummyEvents(clientFactory, streamName, totalEvents, 0);
+    }
+
+    // End region utils
 }
```
Comment (shrids) [inline test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndTruncationTest.java]: Nitpick: add a new line

**Comment by RaulGracia** [comment]
https://github.com/pravega/pravega/issues/2546

File: test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndTruncationTest.java [hunk 2]
```diff
@@ -152,11 +166,292 @@ public void testTruncation() throws Exception {
...
+                event = reader.readNextEvent(timeout);
+                Exceptions.handleInterrupted(() -> Thread.sleep(interReadWait));
+                if (event.getEvent() != null) {
+                    validEvents++;
+                }
+            } while ((event.getEvent() != null || event.isCheckpoint()) && validEvents < limit);
+
+            reader.close();
+        } catch (ReinitializationRequiredException e) {
+            e.printStackTrace();
```
Comment (fpj) [inline test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndTruncationTest.java]: Do not print the stack trace, log it using slf4j. We already have the lombok annotation for this class.

Comment (RaulGracia) [reply test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndTruncationTest.java]: Missed that one. Now `ReinitializationRequiredException` are propagated to the caller.

[the remark this answers is not in the document]
Comment (RaulGracia) [reply test/integration/src/test/java/io/pravega/test/integration/endtoendtest/EndToEndTruncationTest.java]: Now `ReinitializationRequiredException` (and other exceptions) are propagated to the caller.

File: test/system/src/test/java/io/pravega/test/system/OffsetTruncationTest.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+            do {
+                event = reader.readNextEvent(1000);
+                if (event.getEvent() != null) {
+                    validEvents++;
+                }
+            } while ((event.getEvent() != null || event.isCheckpoint()) && validEvents < limit);
+
+            reader.close();
+        } catch (ReinitializationRequiredException e) {
+            log.error("Exception while reading event: ", e);
```
[the remark this answers is not in the document]
Comment (RaulGracia) [reply test/system/src/test/java/io/pravega/test/system/OffsetTruncationTest.java]: Exceptions are now propagated to the caller.

**Pull request merged**